cmds/tail: added tail command

insomniacslk opened the pull request:
Added new command: tail.

This is a very simple implementation, it currently supports only the `-n` (number of lines) option, and
adds the scaffolding for the `-f` (follow) option. It can read from seekable and non-seekable files (e.g.
stdin). If stdin is detected, it will try reading from the beginning instead of reading backwards from the
end.

insomniacslk commented:
I've implemented basic unit tests and done a bunch of comparisons with `tail` from coreutils. Far from being comprehensive, but all the basic cases (short files, long files, with and without newline at the end, empty, one line) have been tried

rminnich commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,243 @@
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile os.File, writer io.Writer, numLines int) error {
+	if numLines < 0 {
```
duplicate code between two functions. can you test in the caller?

insomniacslk replied:
I can, but the main reason I didn't do that is that I wanted this check in the function that may fail if this condition is not met. This is more valid for reusable code, which is probably not the case here, so yeah, I can move the check outside.
An alternative would be to provide a `Config` object instead of `numLines` and have another function that runs sanity checks on the Config. This would be future-proof if I add other tail features like "retry until the file exists" and such. What do you think? Better to start simple and add this as a TODO?

rminnich commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,243 @@
+
+func Tail(inFile os.File, writer io.Writer, follow bool, numLines int) error {
+	var err error
+	if inFile == *os.Stdin {
```
So in general this test is not sufficient. What if they give the name /dev/fd/0? That's stdin. What about all the other character devices? 

 If you really want to do this you need to stat the file, see if it's a block device, then call your backwards. 

Or, simpler, see if seeking from the end of tne file works and, if so, read backwards. Maybe a thing is seekable even if it's not a block device.

insomniacslk replied:
very valid point. I will change this to try to read backwards and fall back to forward reading if that fails

insomniacslk commented:
I will add the missing files, it seems that .gitignore is hitting the directory with the two sample files that I'm using in tests.

On the other failure: `Blksize` has different integer types depending on the architecture. The way I mean to use it is to return a block size that is "good enough" to avoid too many read syscalls, so even if the value is not accurate this is fine. I will downcast it to `int32` since it's the smallest supported integer (even if it's signed)

rminnich commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,243 @@
+
+// getBlockSize returns the block size for the disk where the file pointed by
+// the reader resides
+func getBlockSize(f os.File) (int64, error) {
```
I am not sure, but is this level of optimization really needed?

rminnich commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return err
+	}
+	// read block by block backwards until `numLines` lines are found
```
why all this seeking? Why not just pread? Then it gets pretty easy. First off, most lines are short. A good heuristic is to take the # lines you want, multiply by 81, then do a pread for the last 81*n bytes of the file, split on \n, and see if you have enough lines. I think the blocksize test is way more than you need to do.

Also, for fun, write a simple Go program that traverses our entire disk, and get a histogram of line lengths -- I'll bet it's under 16 chars per line ...

insomniacslk replied:
very good point, I didn't consider `pread` (maybe just because I rarely use it). Will rewrite using it, thanks!

rminnich replied:
sorry about that, I meant ReadAt.

insomniacslk commented:
* Addressed feedback on duplicate code
* Added `TailConfig` object
* Better checks for non-seekable files

Will use `pread` in the next commit

insomniacslk commented:
Note: using `Pread` means using the `syscall` package, too, as this function is not exported in the file interface. It also means getting the file descriptor from the `File` object (which is architecture-dependent). 
@rminnich do you still think it's worth using `syscall.Pread` instead of `Seek` + `Read`? (I'm not arguing that we shouldn't, just genuinely asking to find the best balance)

insomniacslk commented:
my previous comment is pointless, since this command will always and only run on a unix environment, scratch that. However `pfd` is unexported and there's no getter method, so I would have to use `syscall.Open` for that, toot

insomniacslk commented:
As suggested by @hugelgupf I can use `ReadAt` (that wraps around `pread`)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+	}
+	// write the requested lines to the writer
+	_, err = writer.Write(data)
+	if err != nil {
```
choose the smallest scope you can whenever possible:

```go
if _, err := writer.Write(data); err != nil {
```

this way, the Write error does not override the existing `err`, which can be confusing.

insomniacslk replied:
ack

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+		return err
+	}
+	// reposition the stream at the end of what we've read
+	_, err = inFile.Seek(lastPos, os.SEEK_SET)
```
Why, and why ignore the error?

insomniacslk replied:
because in the function doc I guarantee that if there is no error, the file pointer will be at the end of the file (or, better said, pointing to the first byte after the printed lines). This will be used when follow-mode will be implemented, and will be just matter of keep reading from where the file is already positioned.

On "why ignore the error", that's a mistake (now fixed locally)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+			if err == io.EOF {
+				break
+			}
+			if err != io.ErrUnexpectedEOF {
```
So do nothing in the io.ErrUnexpectedEOF case? That seems strange.

insomniacslk replied:
if we get `ErrUnexpectedEOF` there was a short read, i.e. it could read something but less than `blkSize` (which is quite normal). So I just go ahead and take `n` bytes (as returned by `ReadAll`) and let the next look get an EOF (and thus read 0 bytes)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile os.File, writer io.Writer, numLines int) error {
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
```
block size seems like a premature optimization, but I see no harm in it either?

insomniacslk replied:
that's what I thought being the most efficient approach in terms of syscalls, but I will instead use what suggested by @rminnich (i.e. blocks of size `81 * numLines`)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+		// (no further than numLines)
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
+		foundNewLines := bytes.Count(slice, []byte{'\n'})
```
You're always recounting `slice`, you could keep track of the existing number and just count `buf[:n]`? Then, you could potentially save an allocation by just saying `slice = append(slice[m:], buf[:n])` at the end.

insomniacslk replied:
yeah, I should have added a comment on making this more efficient and don't re-count every time. OK, time to do that :)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+			newLinesToSkip = foundNewLines - numLines
+		}
+		// if there are more than `numLines` lines, skip them
+		for idx := 0; idx < newLinesToSkip; idx++ {
```
I feel like this could be solved by using `IndexFunc`: https://play.golang.org/p/dJdHBHwsjbd

```go
m := bytes.IndexFunc(slice, NthByte(newLinesToSkip, '\n'))
if m == -1 {
   return fmt.Errorf("less bytes than expected were found")
}
slice = append(slice[m:], buf[:n])
```

insomniacslk replied:
nice, I didn't know `IndexFunc`, thanks!

insomniacslk replied:
I had a second thought - while I like the elegance of your solution, maybe it's not ideal this dance of converting to utf8 and making a function call for every rune (not byte). I'd rather stick with the original, yet less elegant code, unless you have a stronger reason not to

hugelgupf replied:
why is it not ideal? utf8 will never give you a rune like "\xFF\xA" if that's what you're worried about. if there's a newline somewhere, it'll be a "\xA" rune.

insomniacslk replied:
I was afraid of doing unnecessary conversions around, but as clarified with Ron this is not the case for Go. I will use `IndexFunc` :)

hugelgupf commented on cmds/tail/tail_test.go:
```diff
@@ -0,0 +1,68 @@
+	if err != nil {
+		t.Error(err)
+	}
+	output := bytes.NewBuffer([]byte{})
```
```go
output := &bytes.Buffer{}
```

is sufficient, bytes.Buffer will allocate bytes as needed.

insomniacslk replied:
ack

hugelgupf commented on cmds/tail/tail_test.go:
```diff
@@ -0,0 +1,68 @@
+		t.Fatal(err)
+	}
+	if !bytes.Equal(expected, got[:n]) {
+		t.Fatalf("Invalid result reading backwards. Expected %v, got %v", string(expected), string(got[:n]))
```
This is super nit-picky, but it's customary in Go to put what you got before what you expected (got before want): https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

insomniacslk replied:
Fixed. Good to know for the future, I've always used the other approach (`want` before `got`)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+	return nil
+}
+
+func Tail(inFile os.File, writer io.Writer, config TailConfig) error {
```
Comment? 

```go
// Tail ...
```

insomniacslk replied:
added (after `golint`'s complaints too :) )

bkochendorfer commented on cmds/tail/tail_test.go:
```diff
@@ -0,0 +1,68 @@
+		t.Error(err)
+	}
+	output := bytes.NewBuffer([]byte{})
+	err = readLastLinesBackwards(*input, output, 3)
```
I think you want to test `readLastLinesFromBeginning` here

insomniacslk replied:
definitely. Same as below

bkochendorfer commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+	if retryFromBeginning {
+		err = readLastLinesFromBeginning(inFile, writer, config.numLines)
+	}
+	if config.follow {
```
Do you want to perform the tail and then throw an error about `-f` not being implemented or throw an error up front? As is this will perform the tail then throw an error. I'm not sure which is more pleasant for someone using `tail`.

insomniacslk replied:
yeah, it's a better user experience to bail out immediately, I've fixed this

bkochendorfer commented on cmds/tail/tail_test.go:
```diff
@@ -0,0 +1,68 @@
+		t.Fatal(err)
+	}
+	if !bytes.Equal(expected, got[:n]) {
+		t.Fatalf("Invalid result reading backwards. Expected %v, got %v", string(expected), string(got[:n]))
```
Update the messaging here to indicate you are trying to read from the beginning.

insomniacslk commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+// size, but a value that is good enough to minimize the number of `read` syscalls
+func getBlockSize(f os.File) (int64, error) {
+	var stat syscall.Stat_t
+	err := syscall.Stat(f.Name(), &stat)
```
Thanks. I am removing this function anyway as per @rminnich 's advice, and I will use `81 * numLines` as block size

insomniacslk commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+func getBlockSize(f os.File) (int64, error) {
+	var stat syscall.Stat_t
+	err := syscall.Stat(f.Name(), &stat)
+	if err != nil {
```
same as above, but good advice anyway :)

insomniacslk commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile os.File, writer io.Writer, numLines int) error {
```
done

insomniacslk replied:
Shall I just let the function error out if the passed `os.File` is `nil`? (i.e. if `inFile` is `nil`, `Seek` will return an error but won't `panic()`). Otherwise I can add a check, which would be duplicated in the two functions.

insomniacslk commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,261 @@
+	if err != nil {
+		// if it failed because it couldn't seek, mark it for retry reading from
+		// the beginning
+		if pathErr, ok := err.(*os.PathError); ok && pathErr.Err.Error() == "illegal seek" {
```
ack. The string comparison was too hacky

insomniacslk commented:
I should have addressed all the feedback, except:
* using `ReadAt` (will do after office hours)
* not counting every newline again with every new block
* use `IndexFunc` to reslice after skipping the newlines in excess

insomniacslk commented:
@hugelgupf / @rminnich / @bkochendorfer , I should have addressed all the feedback, can you please have another look?

insomniacslk commented:
Note that I am still using two `Seek` calls to position the stream at the end of the file. The first one could be replaced by `Stat` to get the file size, but I believe that there's no performance or readability gain as it's one syscall in both cases. The latter is for leaving the stream positioned at the end of the file as defined by the function interface - this will be used for the implementation of the follow-mode so the reader is already positioned correctly.

rminnich commented:
Apropos this comment: "making a function call for every rune (not byte)"

Don't overthink this. The Go compiler series is derived from the plan 9 compiler tools (though rewritten in Go, now) and the plan 9 tools did inlining *in the linker* (as does clang nowadays, I am told). It's almost never a good idea to assume you know that the code you write is the code that is run, of even generated. Back in the day, yes, but I've had lots of meetings explaining to people that where in the function you declare the variable no longer really makes that big a difference. Just something to keep in mind.

insomniacslk commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,254 @@
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile *os.File, writer io.Writer, numLines int) error {
```
that's what I tried initially (making it an `io.Reader` - before using `ReadAt`), but those interfaces do not support `Seek`

rminnich replied:
io.ReaderAt, not io.Reader.

insomniacslk replied:
yeah, I tried both, and they both have the same limit on `Seek`

insomniacslk replied:
what I meant is that I originally tried it as an `io.Reader`, then had to switch to `os.File` because of `Seek`. After your advice I tried `io.ReaderAt` too, which presents the same issue

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+func readLastLinesBackwards(inFile *os.File, writer io.Writer, numLines int) error {
+	blkSize := getBlockSize(numLines)
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
```
I think this invariant of always being at the end of the file makes this function a little complicated. if you just pass back the lastPos to the caller, the caller can decide to use it to seek (or ReadAt, or whatever) when -f is passed.

insomniacslk replied:
I agree. In fact my original implementation was returning `(int64, error)` instead of just `error`. I changed that because the function `readLastLinesFromBeginning` cannot return a position. In C I would have used `ftell` but Go does not expose it. Searching on Google it seems that the common workaround is to `Seek(0, os.SEEK_CUR)`, which cannot be done in `readLastLinesFromBeginning`, as it's used for non-seekable files.

If I don't enforce this behaviour, I have to make the caller (`Tail()`) more cumbersome, because it will call the two functions that do the same thing (yet in different ways) with two different signatures, i.e.:

```
err = readLastLinesBackwards(inFile, writer, config.numLines)
...
if retryFromBeginning {
    err = readLastLinesFromBeginning(inFile, writer, config.numLines) 
}
```

I don't have strong opinions on whether it's better to have two discrepant signatures or to use a `io.ReaderAt` instead of `os.File`. Which is why it's good to discuss it here :)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	// find the start of the n-th to last line
+	slice := readData
```
I preferred it a bit when this was still inside the loop. That way, no more than 2*chunkSize is ever in the buffer, rather than the entire file.

hugelgupf replied:
Oh, nevermind, I'm sorry. I thought this was the other one.

insomniacslk replied:
np, I just realized the same :)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+	if len(slice) != 0 {
+		if slice[len(slice)-1] == '\n' {
+			// don't consider the last new line for the line count
+			slice = slice[:len(slice)-1]
```
bytes.TrimRight(slice, "\n")

insomniacslk replied:
Looking at the docs, `TrimRight` would remove more than one newline, if more than one is found, which is not what I want (i.e. removing only the last one, if present)

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+		}
+		pos -= thisChunkSize
+		n, err := inFile.ReadAt(buf, pos)
+		if err != nil {
```
could be replaced with?

```go
if err != nil && err != io.EOF {
  return err
}
```

`io.ReaderAt` implementations must return an error if it's a short read, so `n != 0` implies `err != nil`.

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	// write the requested lines to the writer
+	if _, err = writer.Write(data); err != nil {
```
if you do move the seek out, do this:

```go
_, err = writer.Write(data)
return err
```
  
or if you keep the seek, do it to the call below.

insomniacslk replied:
ack. For now I've done it in the call below. I'll wait for the discussion at the comment above to eventually do it above, https://github.com/u-root/u-root/pull/548#discussion_r159772382

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+// TailConfig is a configuration object for the Tail function
+type TailConfig struct {
+	follow   bool // enable follow-mode (-f)
+	numLines int  // specifies the number of lines to print (-n)
```
why isn't this a uint?

insomniacslk replied:
I did it to simplify the code for all the comparisons and other typed operations, and because there is a check for non-negativeness before passing it (even if I don't feel comfortable with checking values in outer functions that are needed in inner functions). I can go back to `uint` here though

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+	// if reading backwards failed because the file is not seekable,
+	// retry from the beginning
+	if retryFromBeginning {
+		err = readLastLinesFromBeginning(inFile, writer, config.numLines)
```
unused err?

insomniacslk replied:
yes. Fixed in the next commit

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,259 @@
+		return fmt.Errorf("follow-mode not implemented yet")
+	}
+	var (
+		err                error
```
it's customary in Go not to declare variables until you actually need them. e.g., if you remove the error declaration here and just use

```go
err := readLastLinesBackwards(...
```
that'll do your error decl for you.

insomniacslk replied:
Absolutely right. This is a leftover of the original code when `readLastLinesBackwards` returned both int64 and error. Fixed in the next commit

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,251 @@
+			break
+		}
+	}
+	// find the start of the n-th to last line
```
Maybe this could be in its own function? There are a lot of slices going around here.

```go
func lastNLines(buf []byte, n int) []byte {
  
}
```

insomniacslk replied:
fixed in the next commit

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,251 @@
+func readLastLinesFromBeginning(inFile *os.File, writer io.Writer, numLines uint) error {
+	blkSize := getBlockSize(numLines)
+	// read block by block until EOF and store a reference to the last lines
+	buf := make([]byte, blkSize) // will hold each block of read data
```
I think the comment here is unnecessary. also, Go doesn't normally do comments after code, you'd have the comment on its own line.

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,251 @@
+		// (no further than numLines)
+		foundLines += uint(bytes.Count(buf[:n], []byte{'\n'}))
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
```
I feel like the code here is another candidate for just calling a `lastNLines` function

insomniacslk replied:
definitely. The code is much shorter now

hugelgupf commented on cmds/tail/tail.go:
```diff
@@ -0,0 +1,251 @@
+
+// TailConfig is a configuration object for the Tail function
+type TailConfig struct {
+	follow   bool // enable follow-mode (-f)
```
Comments on their own line:

```go
type TailConfig struct {
  // follow is whether follow-mode is enabled (-f).
  follow bool
  
  // numLines specifies the number of lines to print (-n).
  numLines uint
}
```

insomniacslk replied:
ack

insomniacslk commented:
Thank you for giving me very good advice throughout this review :)
